Add shell api to set default for windows data

chunhtai opened the pull request:
This pr enables embedding to set default values for window data.

It does following things

1. Make WindowData a public class in runtime controller
2. Add windowData a construction parameter for RuntimeController, Engine, and Shell.
3. If there is no windowData passed in to Shell:Create, it will use the default version of WindowData.
4, Makes Android and IOS pass in windowData with their own defaults


relevant issue: https://github.com/flutter/flutter/issues/45270

chunhtai commented on runtime/runtime_controller.cc:
```diff
@@ -14,37 +14,6 @@
 
 namespace flutter {
 
-RuntimeController::RuntimeController(
```
RuntimeController constructor will always need window data.

chunhtai commented:
cc @xster

chinmaygarde commented:
Can you elaborate on the use case? The default frame size is zero and a frame wont get rendered by the rasterizer at that size. So the first frame that ends up on screen will already be the result of the setter to the viewport metrics.

chunhtai commented:
> Can you elaborate on the use case? The default frame size is zero and a frame wont get rendered by the rasterizer at that size. So the first frame that ends up on screen will already be the result of the setter to the viewport metrics.

The windows data here is the default value, those are to be set before user setting and viewport metric is flushed to windows.dart. In the most case, we don't care about the default value because those setting will be flushed during viewcontroller/activity creating. In the add to app scenario, the view controller or activity may not exist when the framework app is running, and the app is running with the default window data. For now, this pr only set applifecycle default value to be detached for android and ios. This pr will open a lot of opportunities to set those default so that the app can warm up in a more reasonable way in the case of missing viewcontroller or activity.

chunhtai commented:
This is also a refactoring. The existing test should cover it. Except in the case of non ios and android platform it uses the 'default' window data in runtime_controller.h. I do not know how to test it. I would imagine the test would be using shell:create to create the shell and verify window.dart has the correct value. I don't know if this test is possible. @chinmaygarde do you have any suggestion?

chunhtai commented:
This requires some more love

cbracken commented:
@chinmaygarde this is ready for re-review.

xster commented:
Also add @gaaclarke

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
```diff
@@ -253,4 +261,10 @@ - (void)setPersistentIsolateData:(NSData*)data {
+#pragma mark - windowData utilities
+
+- (void)initializesWindowData {
+  _windowData.lifecycle_state = std::string([@"AppLifecycleState.detached" UTF8String]);
```
This can just be `"AppLifecycleState.detached"`.

gaaclarke commented on runtime/runtime_controller.h:
```diff
@@ -26,6 +26,26 @@ class RuntimeDelegate;
+struct WindowData {
+  WindowData();
+
+  WindowData(const WindowData& other);
```
Why do we need a custom copy constructor?  If we need one we should probably do something about operator= as well.

gaaclarke commented on shell/platform/android/android_shell_holder.cc:
```diff
@@ -101,8 +101,12 @@ AndroidShellHolder::AndroidShellHolder(
   );
 
+  flutter::WindowData window_data;
+  InitializesWindowData(window_data);
```
Instead of using a mutating function, why not just create the WindowData?

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
```diff
@@ -148,6 +148,7 @@
 
 @implementation FlutterDartProject {
   flutter::Settings _settings;
+  flutter::WindowData _windowData;
```
I don't see why this needs to be an instance variable.

chunhtai commented:
> I would like to see the change that fixes the issue in the PR as well.

Previously the it will set the lifecylce enum to be detached by default for all platforms, and expects the platform to update the lifecycle enum when they are ready.
This pr fixes this issue by setting the default to detached for ios and android where both platform will update the enum, and set the default to null for all the other platforms who currently do not opt in to handle lifecycle. I provide the api for each platform to set the default once they opt in to handle lifecycle enum.

I also addressed all comments, this is ready for review

gaaclarke commented on runtime/runtime_controller.cc:
```diff
@@ -57,7 +26,7 @@ RuntimeController::RuntimeController(
     std::string p_advisory_script_entrypoint,
     const std::function<void(int64_t)>& idle_notification_callback,
-    WindowData p_window_data,
+    const WindowData p_window_data,
```
Ideally this should be a const ref:
`const WindowData& p_window_data`

gaaclarke commented on runtime/runtime_controller.h:
```diff
@@ -40,6 +41,7 @@ class RuntimeController final : public WindowClient {
       std::string advisory_script_uri,
       std::string advisory_script_entrypoint,
       const std::function<void(int64_t)>& idle_notification_callback,
+      const WindowData data,
```
Const ref here too.

gaaclarke commented on runtime/window_data.h:
```diff
@@ -0,0 +1,48 @@
+/// The struct of platform-specific data used for initializing ui.Window.
+///
+/// framework may request data from ui.Window before platform is properly
+/// configured. Uses this struct to set the desired default value for ui.Window
```
Who "uses this struct"?  I'd mention Engine and add a link to it.

cbracken commented on runtime/window_data.h:
```diff
@@ -0,0 +1,49 @@
+///
+///  * flutter::Shell::Create, which takes a window_data to initialize the
+///  ui.Window
+///    attached to it.
```
nit: needs re-formatting.

cbracken commented on shell/common/shell.h:
```diff
@@ -126,6 +126,44 @@ class Shell final : public PlatformView::Delegate,
+  ///             Unlike the simpler variant of this factory method, this method
+  ///             allows for specification of window data. If this is the first
+  ///             instance of a shell in the process, this
+  ///             call also bootstraps the Dart VM.
```
nit: line wrapping.

cbracken commented on shell/platform/android/android_shell_holder.cc:
```diff
@@ -140,6 +141,12 @@ AndroidShellHolder::~AndroidShellHolder() {
   FML_CHECK(pthread_key_delete(thread_destruct_key_) == 0);
 }
 
+WindowData AndroidShellHolder::getDefaultWindowData() {
```
Given that this doesn't rely on object state at all, this could be a static function local to the TU.

cbracken replied:
(Either way this function should start with a capital though)

cbracken replied:
On third thought, if you don't expect this to change significantly in the future or be called from a ton of places, I'd probably just inline the variable declaration and field initialisation at the usage site above.

chunhtai replied:
I separated this out to a method for readability. We might want to fill in more default value for add 2 app, for example, viewport metrics.
